xop.go

```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
 	}
 }
 
-func getFieldByName(element reflect.Value, name string) (reflect.Value, error) {
-	for i := 0; i < element.NumField(); i++ {
-		rTypeField := element.Type().Field(i)
-		rValueField := element.Field(i)
+func getFieldFromPath(val reflect.Value, path []string) (reflect.Value, error) {
+	val = unwrapValue(val)
 
-		if rTypeField.Name == name {
-			return rValueField, nil
-		}
-	}
-	return reflect.Value{}, errFieldNotFound
-}
-
-// This assumes, if it encounters an array field, that it will be filling in the first field and continuing
-// TODO: support arbitrary array size.
-func getFieldByXMLTagPrefix(element reflect.Value, prefix string) (reflect.Value, error) {
-	if element.Kind() == reflect.Interface || element.Kind() == reflect.Ptr {
-		element = element.Elem()
-	}
-	if element.Kind() == reflect.Array || element.Kind() == reflect.Slice {
-		return getFieldByXMLTagPrefix(element.Index(0), prefix)
-	}
-	if element.Kind() != reflect.Struct {
+	// val must be a struct and path must have length > 0
+	if val.Type().Kind() != reflect.Struct || len(path) == 0 {
 		return reflect.Value{}, errFieldNotFound
 	}
 
-	for i := 0; i < element.NumField(); i++ {
-		rTypeField := element.Type().Field(i)
-		rValueField := element.Field(i)
+	// search the struct fields with the path
+	for i := 0; i < val.NumField(); i++ {
+		typeField := val.Type().Field(i)
+		valueField := val.Field(i)
+		tag := typeField.Tag.Get("xml")
+
+		// skip the XMLName field
+		if typeField.Name == xmlName {
+			continue
+		}
 
-		if tagValue, tagOk := rTypeField.Tag.Lookup("xml"); tagOk {
-			if !strings.HasPrefix(tagValue, prefix) {
-				continue
+		// skip omitted fields
+		if tag == "-" {
+			continue
+		}
+
+		// unwrap the value
+		valueField = unwrapValue(valueField)
+
+		// check if the value was unwrapped completely
+		if valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Ptr {
```

Comment:
I think you're checking whether it's a Slice twice in this conditional? That seems unnecessary?

Reply:
Did you mean to check for Slice, Array and Ptr?

Reply:
yes, I meant to check for Slice, Array and Ptr